### PR TITLE
Implement Dirichlet-Multinomial distribution

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #1707 

## Tested

- added tests to test_conjugate.py

### CODE REVIEW

File: pyro/distributions/conjugate.py [hunk 1]
```diff
@@ -87,6 +93,70 @@ def enumerate_support(self, expand=True):
...
+        return Binomial(self.total_count, probs).sample()
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        n = self.total_count
+        alpha = self.concentration
+        alpha_sum = self.concentration.sum(-1)
+        return (_log_factorial(n) + torch.lgamma(alpha_sum) - torch.lgamma(n + alpha_sum) +
+                (torch.lgamma(value + alpha) - _log_factorial(value) - torch.lgamma(alpha)).sum(-1))
```
Comment (fehiepsi) [inline pyro/distributions/conjugate.py]: @fritzo just a side note (for us to keep in mind if it helps for performance): when `value` is sparse, we can avoid computing `torch.lgamma(value + alpha) - _log_factorial(value) - torch.lgamma(alpha)` at zero-value indices of `value`.  However, to do so, we have to flatten `value` and `alpha` (to deal with batch_shape), select nonzero indices, compute the result, then transform back to match the batch_shape. I will address this note in the future when facing this distribution in a real case (if it helps).

File: pyro/distributions/conjugate.py [hunk 2]
```diff
@@ -87,6 +93,70 @@ def enumerate_support(self, expand=True):
...
+        batch_shape = torch.Size(batch_shape)
+        new._dirichlet = self._dirichlet.expand(batch_shape)
+        new.total_count = self.total_count.expand(batch_shape)
+        super(DirichletMultinomial, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def sample(self, sample_shape=()):
+        probs = self._dirichlet.sample(sample_shape)
+        return Binomial(self.total_count, probs).sample()
```
Comment (neerajprad) [inline pyro/distributions/conjugate.py]: This should be Multinomial.

**Comment by fritzo** [comment]
@neerajprad thanks for the detailed review. I've added a new test to conftest.py and fixed a bug in `Multinomial.support` so that these tests pass.

File: pyro/distributions/conjugate.py [hunk 3]
```diff
@@ -87,6 +93,75 @@ def enumerate_support(self, expand=True):
...
+        super(DirichletMultinomial, new).__init__(
+            new._dirichlet.batch_shape, new._dirichlet.event_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def sample(self, sample_shape=()):
+        probs = self._dirichlet.sample(sample_shape)
+        total_count = int(self.total_count.max())
+        if not self.total_count.min() == total_count:
+            raise NotImplementedError("Inhomogeneous total count not supported by `enumerate_support`.")
```
Comment (neerajprad) [inline pyro/distributions/conjugate.py]: enumerate_support --> sample

File: pyro/distributions/torch_patch.py [hunk 4]
```diff
@@ -59,6 +59,16 @@ def _torch_linspace(*args, **kwargs):
     return ret
 
 
+# Fixes a shape error in Multinomial.support with inhomogeneous .total_count
+@patch_dependency('torch.distributions.Multinomial.support')
+@torch.distributions.constraints.dependent_property
+def _Multinomial_support(self):
```
Comment (neerajprad) [inline pyro/distributions/torch_patch.py]: Nice catch, I completely forgot the Multinomial upstream only supports integer total count.

**Pull request merged**